### PR TITLE
add `get_pvol`

### PR DESCRIPTION
**Pull request by bart1**
This pull request implements downloads of polar volumes for `dk`, `nl`, `de`, `cz`, `ee` & `fi`

It implements some basic testing to see if it works, however as tests are dependent on authentication and stable connections not all tests will always run.

I have kept most dependencies as suggests. One I doubted about was `bioRad` as it adds a lot of other dependencies. But as it is core to pvol I kept is in import

### CODE REVIEW

**Comment by bart1** [comment]
Maybe a few more notes on some design:
- I try to consistently set the user agent so requests through the package can be tracked. Would it be useful to add the package version number or commit here? If so how to keep that up to date.
- I now class each error as `"getRad_error_...."` we could also go for `c("getRad_error_....","getRad_error")` more like `httr2`
- currently `time` should be a `POSIXct` rounded to 5 minutes we might consider also allowing for intervals and multiple timestamps. The individual functions now only download one file at a time and do not parallelize requests.

**Comment by peterdesmet** [comment]
Can you list your design notes as issues? Easier to answer and track.

Longer term we’ll have to be careful not to introduce circular dependencies between this and bioRad

**Comment by bart1** [comment]
> Can you list your design notes as issues? Easier to answer and track.
> 

See #2, #3 & #4

**Comment by bart1** [comment]
I have also added a bit of examples to the readme

**Comment by bart1** [comment]
@peterdesmet I think this is quite a coherent set of functions that can be added

**Comment by PietrH** [comment]
I'm currently failing a few tests locally, looking into it: 

```r
── Failed tests ────────────────────────────────────────────────────────────
Failure (test-get_pvol.R:32:3): multiple timestamps work
lapply(pvl, function(x) x$datetime) (`actual`) not identical to as.list(t) (`expected`).

`attr(actual[[1]], 'tzone')` is a character vector ('UTC')
`attr(expected[[1]], 'tzone')` is absent

`attr(actual[[2]], 'tzone')` is a character vector ('UTC')
`attr(expected[[2]], 'tzone')` is absent

Failure (test-get_pvol.R:40:3): multiple timestamps and radars work
lapply(pvl, function(x) x$datetime) (`actual`) not identical to as.list(rep(t, each = 2)) (`expected`).

`attr(actual[[1]], 'tzone')` is a character vector ('UTC')
`attr(expected[[1]], 'tzone')` is absent

`attr(actual[[2]], 'tzone')` is a character vector ('UTC')
`attr(expected[[2]], 'tzone')` is absent

`attr(actual[[3]], 'tzone')` is a character vector ('UTC')
`attr(expected[[3]], 'tzone')` is absent

`attr(actual[[4]], 'tzone')` is a character vector ('UTC')
`attr(expected[[4]], 'tzone')` is absent
```

**Comment by bart1** [comment]
> I'm currently failing a few tests locally, looking into it:
> 
> ```r
> ── Failed tests ────────────────────────────────────────────────────────────
> Failure (test-get_pvol.R:32:3): multiple timestamps work
> lapply(pvl, function(x) x$datetime) (`actual`) not identical to as.list(t) (`expected`).
> 
> `attr(actual[[1]], 'tzone')` is a character vector ('UTC')
> `attr(expected[[1]], 'tzone')` is absent
> 
> `attr(actual[[2]], 'tzone')` is a character vector ('UTC')
> `attr(expected[[2]], 'tzone')` is absent
> 
> Failure (test-get_pvol.R:40:3): multiple timestamps and radars work
> lapply(pvl, function(x) x$datetime) (`actual`) not identical to as.list(rep(t, each = 2)) (`expected`).
> 
> `attr(actual[[1]], 'tzone')` is a character vector ('UTC')
> `attr(expected[[1]], 'tzone')` is absent
> 
> `attr(actual[[2]], 'tzone')` is a character vector ('UTC')
> `attr(expected[[2]], 'tzone')` is absent
> 
> `attr(actual[[3]], 'tzone')` is a character vector ('UTC')
> `attr(expected[[3]], 'tzone')` is absent
> 
> `attr(actual[[4]], 'tzone')` is a character vector ('UTC')
> `attr(expected[[4]], 'tzone')` is absent
> ```

Hmm interesting it looks like this relates to hoe expectations are generated , for me all these example have the `tzone` attribute. Could you check on your computer:
``` r
as.POSIXct(Sys.Date()) |> dput()
#> structure(1730073600, class = c("POSIXct", "POSIXt"), tzone = "UTC")
(c(-300, -900) + as.POSIXct(Sys.Date())) |> dput()
#> structure(c(1730073300, 1730072700), class = c("POSIXct", "POSIXt"
#> ), tzone = "UTC")
rep(c(-300, -900) + as.POSIXct(Sys.Date()), each=TRUE) |> dput()
#> structure(c(1730073300, 1730072700), class = c("POSIXct", "POSIXt"
#> ), tzone = "UTC")

sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.4.0 (2024-04-24)
#>  os       Ubuntu 22.04.4 LTS
#>  system   x86_64, linux-gnu
#>  ui       X11
#>  language (EN)
#>  collate  en_US.UTF-8
#>  ctype    en_US.UTF-8
#>  tz       Europe/Amsterdam
#>  date     2024-10-28
#>  pandoc   3.1.11 @ /usr/lib/rstudio/resources/app/bin/quarto/bin/tools/x86_64/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version date (UTC) lib source
#>  cli           3.6.3   2024-06-21 [1] CRAN (R 4.4.0)
#>  digest        0.6.35  2024-03-11 [1] CRAN (R 4.4.0)
#>  evaluate      0.23    2023-11-01 [1] CRAN (R 4.4.0)
#>  fastmap       1.2.0   2024-05-15 [1] CRAN (R 4.4.0)
#>  fs            1.6.4   2024-04-25 [1] CRAN (R 4.4.0)
#>  glue          1.8.0   2024-09-30 [1] CRAN (R 4.4.0)
#>  htmltools     0.5.8.1 2024-04-04 [1] CRAN (R 4.4.0)
#>  knitr         1.46    2024-04-06 [1] CRAN (R 4.4.0)
#>  lifecycle     1.0.4   2023-11-07 [1] CRAN (R 4.4.0)
#>  magrittr      2.0.3   2022-03-30 [1] CRAN (R 4.4.0)
#>  purrr         1.0.2   2023-08-10 [1] CRAN (R 4.4.0)
#>  R.cache       0.16.0  2022-07-21 [1] CRAN (R 4.4.0)
#>  R.methodsS3   1.8.2   2022-06-13 [1] CRAN (R 4.4.0)
#>  R.oo          1.26.0  2024-01-24 [1] CRAN (R 4.4.0)
#>  R.utils       2.12.3  2023-11-18 [1] CRAN (R 4.4.0)
#>  reprex        2.1.0   2024-01-11 [1] CRAN (R 4.4.0)
#>  rlang         1.1.4   2024-06-04 [1] CRAN (R 4.4.0)
#>  rmarkdown     2.26    2024-03-05 [1] CRAN (R 4.4.0)
#>  rstudioapi    0.16.0  2024-03-24 [1] CRAN (R 4.4.0)
#>  sessioninfo   1.2.2   2021-12-06 [1] CRAN (R 4.4.0)
#>  styler        1.10.3  2024-04-07 [1] CRAN (R 4.4.0)
#>  vctrs         0.6.5   2023-12-01 [1] CRAN (R 4.4.0)
#>  withr         3.0.1   2024-07-31 [1] CRAN (R 4.4.0)
#>  xfun          0.43    2024-03-25 [1] CRAN (R 4.4.0)
#>  yaml          2.3.9   2024-07-05 [1] CRAN (R 4.4.0)
#> 
#>  [1] /home/bart/R/x86_64-pc-linux-gnu-library/4.4
#>  [2] /usr/local/lib/R/site-library
#>  [3] /usr/lib/R/site-library
#>  [4] /usr/lib/R/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

**Comment by PietrH** [comment]
Indeed, I'm missing `tzone`
``` r
as.POSIXct(Sys.Date()) |> dput()
#> structure(1730073600, class = c("POSIXct", "POSIXt"))
```

``` r
(c(-300, -900) + as.POSIXct(Sys.Date())) |> dput()
#> structure(c(1730073300, 1730072700), class = c("POSIXct", "POSIXt"
#> ))
```

``` r
rep(c(-300, -900) + as.POSIXct(Sys.Date()), each=TRUE) |> dput()
#> structure(c(1730073300, 1730072700), class = c("POSIXct", "POSIXt"
#> ))
```

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.2.1 (2022-06-23)
#>  os       Ubuntu 22.10
#>  system   x86_64, linux-gnu
#>  ui       X11
#>  language en_US
#>  collate  en_US.UTF-8
#>  ctype    en_US.UTF-8
#>  tz       Europe/Brussels
#>  date     2024-10-28
#>  pandoc   3.1.11 @ /usr/lib/rstudio/resources/app/bin/quarto/bin/tools/x86_64/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version date (UTC) lib source
#>  cli           3.6.3   2024-06-21 [1] CRAN (R 4.2.1)
#>  digest        0.6.36  2024-06-23 [1] CRAN (R 4.2.1)
#>  evaluate      0.24.0  2024-06-10 [1] CRAN (R 4.2.1)
#>  fastmap       1.2.0   2024-05-15 [1] CRAN (R 4.2.1)
#>  fs            1.6.4   2024-04-25 [1] CRAN (R 4.2.1)
#>  glue          1.7.0   2024-01-09 [1] CRAN (R 4.2.1)
#>  htmltools     0.5.8.1 2024-04-04 [1] CRAN (R 4.2.1)
#>  knitr         1.47    2024-05-29 [1] CRAN (R 4.2.1)
#>  lifecycle     1.0.4   2023-11-07 [1] CRAN (R 4.2.1)
#>  magrittr      2.0.3   2022-03-30 [1] CRAN (R 4.2.1)
#>  purrr         1.0.2   2023-08-10 [1] CRAN (R 4.2.1)
#>  R.cache       0.16.0  2022-07-21 [1] CRAN (R 4.2.1)
#>  R.methodsS3   1.8.2   2022-06-13 [1] CRAN (R 4.2.1)
#>  R.oo          1.26.0  2024-01-24 [1] CRAN (R 4.2.1)
#>  R.utils       2.12.3  2023-11-18 [1] CRAN (R 4.2.1)
#>  reprex        2.1.0   2024-01-11 [1] CRAN (R 4.2.1)
#>  rlang         1.1.4   2024-06-04 [1] CRAN (R 4.2.1)
#>  rmarkdown     2.27    2024-05-17 [1] CRAN (R 4.2.1)
#>  rstudioapi    0.16.0  2024-03-24 [1] CRAN (R 4.2.1)
#>  sessioninfo   1.2.2   2021-12-06 [1] CRAN (R 4.2.1)
#>  styler        1.10.3  2024-04-07 [1] CRAN (R 4.2.1)
#>  vctrs         0.6.5   2023-12-01 [1] CRAN (R 4.2.1)
#>  withr         3.0.1   2024-07-31 [1] CRAN (R 4.2.1)
#>  xfun          0.45    2024-06-16 [1] CRAN (R 4.2.1)
#>  yaml          2.3.10  2024-07-26 [1] CRAN (R 4.2.1)
#> 
#>  [1] /home/pieter_huybrechts/R/x86_64-pc-linux-gnu-library/4.2
#>  [2] /usr/local/lib/R/site-library
#>  [3] /usr/lib/R/site-library
#>  [4] /usr/lib/R/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

<sup>Created on 2024-10-28 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by bart1** [comment]
Intersting appearently a change in how `as.POSIXct` works, most likely in R 4.3.0:

```
as.POSIXlt(<Date>) now does apply a tz (time zone) argument, as does as.POSIXct(); partly suggested by Roland Fuß on the R-devel mailing list.
```
https://cran.r-project.org/doc/manuals/r-release/NEWS.html
Maybe it is best if you can just add a tz argument to the expectation and see if it works on your computer?

**Comment by PietrH** [comment]
> > Having considered it, I think we should avoid the `Sys.Date()` call altogether and use `lubridate::ymd_hms()` with an arbitrary date instead. I suppose something like `lubridate::now()` would also work.
> 
> `now` could work but the date needs to be in the last few days as German data is not public for more then 3 days ....

Right, I'll make a suggestion

**Comment by PietrH** [comment]
I created an issue for two identical mentions of checking the timezone formatting with the expectation from the API:  #8

**Comment by bart1** [comment]
I see the tests already covered this by downloading accessing in an other timezone and confirming the resulting date matches

**Comment by PietrH** [comment]
3 tests are skipped for me:

```
• Because no key for Denmark is available in the testing environment (1):
  test-get_pvol_dk.R:2:3
• exists("pvol") is not TRUE (1): test-get_pvol_ee.R:10:3
• Sys.which("KNMI_vol_h5_to_ODIM_h5") == "" is TRUE (1): test-get_pvol_nl.R:3:3
```

- no key: ok
- missing `pvol` object : test might be broken? 
- Looking for an executable, which I do not seem to have: `KNMI_vol_h5_to_ODIM_h5`, this is the same name as the included `.c` script. Am I missing a step?

**Comment by bart1** [comment]
> 
>     * no key: ok
> 
>     * missing `pvol` object : test might be broken?
> 
>     * Looking for an executable, which I do not seem to have: `KNMI_vol_h5_to_ODIM_h5`, this is the same name as the included `.c` script. Am I missing a step?

For the KNMI converter I would suggest users compile themself, converting it to R seems complicated until we know usage picks up. The error message suggests to compile it and gives some instructions

The Estonian api seems frequently over used meaning it frequently fails at downloading. So I for the time being skip the test if the download failed

**Comment by PietrH** [comment]
> > ```
> > * no key: ok
> > 
> > * missing `pvol` object : test might be broken?
> > 
> > * Looking for an executable, which I do not seem to have: `KNMI_vol_h5_to_ODIM_h5`, this is the same name as the included `.c` script. Am I missing a step?
> > ```
> 
> For the KNMI converter I would suggest users compile themself, converting it to R seems complicated until we know usage picks up. The error message suggests to compile it and gives some instructions

The error message I get for the test is : ` Sys.which("KNMI_vol_h5_to_ODIM_h5") == "" is TRUE (1): test-get_pvol_nl.R:3:3`, I suggest adding a comment to the test, and compiling instructions to the Readme. 

> The Estonian api seems frequently over used meaning it frequently fails at downloading. So I for the time being skip the test if the download failed

There might be a better way to go about this than using `  skip_if_not(exists("pvol"))`, maybe add a comment and a custom skip message?

**Comment by bart1** [comment]
> > > ```
> > > * no key: ok
> > > 
> > > * missing `pvol` object : test might be broken?
> > > 
> > > * Looking for an executable, which I do not seem to have: `KNMI_vol_h5_to_ODIM_h5`, this is the same name as the included `.c` script. Am I missing a step?
> > > ```
> > 
> > 
> > For the KNMI converter I would suggest users compile themself, converting it to R seems complicated until we know usage picks up. The error message suggests to compile it and gives some instructions
> 
> The error message I get for the test is : ` Sys.which("KNMI_vol_h5_to_ODIM_h5") == "" is TRUE (1): test-get_pvol_nl.R:3:3`, I suggest adding a comment to the test, and compiling instructions to the Readme.
> 
> > The Estonian api seems frequently over used meaning it frequently fails at downloading. So I for the time being skip the test if the download failed
> 
> There might be a better way to go about this than using ` skip_if_not(exists("pvol"))`, maybe add a comment and a custom skip message?

Sounds both good, I'm however unable to provide compilation instructions for windows (lack of knowledge), currently I can;t recreate the estonia problem so I will see if I can improve the test

**Review by PietrH — CHANGES_REQUESTED** [review]

**Comment by PietrH** [comment]
> > > > ```
> > > > * no key: ok
> > > > 
> > > > * missing `pvol` object : test might be broken?
> > > > 
> > > > * Looking for an executable, which I do not seem to have: `KNMI_vol_h5_to_ODIM_h5`, this is the same name as the included `.c` script. Am I missing a step?
> > > > ```
> > > 
> > > 
> > > For the KNMI converter I would suggest users compile themself, converting it to R seems complicated until we know usage picks up. The error message suggests to compile it and gives some instructions
> > 
> > 
> > The error message I get for the test is : ` Sys.which("KNMI_vol_h5_to_ODIM_h5") == "" is TRUE (1): test-get_pvol_nl.R:3:3`, I suggest adding a comment to the test, and compiling instructions to the Readme.
> > > The Estonian api seems frequently over used meaning it frequently fails at downloading. So I for the time being skip the test if the download failed
> > 
> > 
> > There might be a better way to go about this than using ` skip_if_not(exists("pvol"))`, maybe add a comment and a custom skip message?
> 
> Sounds both good, I'm however unable to provide compilation instructions for windows (lack of knowledge), currently I can;t recreate the estonia problem so I will see if I can improve the test

The CI here on Github is running into the same test skips.

File: R/aaa.R [hunk 1]
```diff
@@ -0,0 +1,18 @@
+.onLoad <- function(libname, pkgname) { # nolint
+
+  op <- options()
+  op.getRad <- list(
+    getRad.key_prefix = "getRad_",
```
[the remark this answers is not in the document]
Comment (bart1) [reply R/aaa.R:5]: Iet also helps if the user needs multiple sets of keys

File: R/aaa.R [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    getRad.user_agent = paste("R package getRad", getNamespaceVersion("getRad"))
```
[the remark this answers is not in the document]
Comment (bart1) [reply R/aaa.R:6]: I think getNamespaceVersion does not get outdated right? It just grabs from the Description/installation

File: R/get_pvol_cz.R [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    glue::glue("http://opendata.chmi.cz/meteorology/weather/radar/sites/{substr(radar,3,5)}/vol_z/hdf5/"),
+    glue::glue("http://opendata.chmi.cz/meteorology/weather/radar/sites/{substr(radar,3,5)}/vol_u/hdf5/"),
+    glue::glue("http://opendata.chmi.cz/meteorology/weather/radar/sites/{substr(radar,3,5)}/vol_v/hdf5/"),
+    glue::glue("http://opendata.chmi.cz/meteorology/weather/radar/sites/{substr(radar,3,5)}/vol_w/hdf5/"),
+    glue::glue("http://opendata.chmi.cz/meteorology/weather/radar/sites/{substr(radar,3,5)}/vol_zdr/hdf5/"),
+    glue::glue("http://opendata.chmi.cz/meteorology/weather/radar/sites/{substr(radar,3,5)}/vol_rhohv/hdf5/"),
+    glue::glue("http://opendata.chmi.cz/meteorology/weather/radar/sites/{substr(radar,3,5)}/vol_phidp/hdf5/")
+  )
+  rlang::check_installed(
+    c("lubridate", "tidyr", "xml2", "rhdf5"),
```
[the remark this answers is not in the document]
Comment (bart1) [reply R/get_pvol_cz.R:16]: Both xml2 and rhdf5 require system libraries to install. I would keep as much as (reasonably) possible in suggest as we might add more functionality. Then you might end up with a package hard to install when you only need simple functionality.

**Comment by PietrH** [comment]
Sure, but we still have to keep description up to date. To help with this
we might want to setup an extra check in CI

File: R/secrets.R [hunk 4]
```diff
@@ -0,0 +1,78 @@
+#' @export
+#' @rdname secret
+get_secret <- function(name) {
+  rlang::check_installed("keyring", "to manage secrets in getRad")
```
[the remark this answers is not in the document]
Comment (bart1) [reply R/secrets.R:4]: Keyring is only needed for `dk` and `nl` so it is really optional.

**Comment by PietrH** [comment]
@peterdesmet, what is your opinion on placing dependencies in Suggests as
much as possible?

File: tests/testthat/test-get_pvol.R [hunk 5]
```diff
@@ -0,0 +1,129 @@
+# Country specific tests are in the respective country specific `get_pvol_`
+# files
+
+test_that("get_pvol radar argument", {
```
[the remark this answers is not in the document]
Comment (bart1) [reply tests/testthat/test-get_pvol.R:4]: I would not do that now as it makes updating and clarifying error messages later more difficult. People wanting to catch the error should really do it on the calss

File: tests/testthat/test-get_pvol_nl.R [hunk 6]
```diff
@@ -0,0 +1,61 @@
+test_that("Pvol for the Netherlands can be downloaded", {
+  testthat::skip_if_offline()
+  testthat::skip_if(Sys.which("KNMI_vol_h5_to_ODIM_h5") == "")
+  withr::local_options(list("keyring_backend" = "env"))
+  # get public key here https://developer.dataplatform.knmi.nl/open-data-api#token
+  withr::local_envvar(
+    list("getRad_nl_api_key" = "eyJvcmciOiI1ZTU1NGUxOTI3NGE5NjAwMDEyYTNlYjEiLCJpZCI6ImE1OGI5NGZmMDY5NDRhZDNhZjFkMDBmNDBmNTQyNjBkIiwiaCI6Im11cm11cjEyOCJ9")
```
[the remark this answers is not in the document]
Comment (bart1) [reply tests/testthat/test-get_pvol_nl.R:7]: These are find as they are public keys from the knmi : https://developer.dataplatform.knmi.nl/open-data-api#token

File: tests/testthat/test-secrets.R [hunk 7]
```diff
@@ -0,0 +1,45 @@
+test_that("get_secret fails correctly", {
```
[the remark this answers is not in the document]
Comment (bart1) [reply tests/testthat/test-secrets.R:1]: I would first check if the messages are clear, for people wanting to capture the errors they should really use the class

File: NEWS.md [hunk 8]
```diff
@@ -0,0 +1,4 @@
+# getRad (development version)
```
[the remark this answers is not in the document]
Comment (bart1) [reply NEWS.md:1]: Sounds good

File: R/get_pvol_de.R [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+
+list_to_scan <- function(x, p) {
+  xx <- x[[1]]
+
+  xx$params <- lapply(x, function(x) x$params[[1]])
+  names(xx$params) <- p
+  xx
+}
+
+read_scan <- function(file, scan = "dataset1",
```
[the remark this answers is not in the document]
Comment (bart1) [reply R/get_pvol_de.R:120]: We can maybe do this now, for the code specific to one country I would keep it separate to keep countries close to each other

**Comment by bart1** [comment]
> Sure, but we still have to keep description up to date. To help with this we might want to setup an extra check in CI
> […](#)
> On Fri, 8 Nov 2024, 16:42 bart1, ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In R/aaa.R <[#1 (comment)](https://github.com/enram/getRad/pull/1#discussion_r1834604450)> : > @@ -0,0 +1,18 @@ +.onLoad <- function(libname, pkgname) { # nolint + + op <- options() + op.getRad <- list( + getRad.key_prefix = "getRad_", + getRad.user_agent = paste("R package getRad", getNamespaceVersion("getRad")) I think getNamespaceVersion does not get outdated right? It just grabs from the Description/installation — Reply to this email directly, view it on GitHub 

I think it will keep roughly up to date, when I checked with peter he did not know a way to add the git hash which could be an improvement but otherwise I think it is good enough for the time being

File: NAMESPACE [hunk 10]
```diff
@@ -1,2 +1,22 @@
...
+importFrom(httr2,req_url)
+importFrom(httr2,req_url_path_append)
+importFrom(httr2,req_url_query)
+importFrom(httr2,req_user_agent)
+importFrom(httr2,request)
+importFrom(httr2,resp_body_html)
+importFrom(httr2,resp_body_json)
+importFrom(httr2,resp_status)
+importFrom(rlang,is_character)
+importFrom(rlang,is_scalar_character)
```
[the remark this answers is not in the document]
Comment (bart1) [reply NAMESPACE:22]: I dont mind name space import but I'm fine with this solution as well

File: R/get_pvol_cz.R [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+  })
+  files_to_get <- data.frame(base = urls) |>
+    dplyr::mutate(file = res) |>
+    tidyr::unnest(file) |>
+    dplyr::filter(file != "../") |>
+    dplyr::mutate(
+      time_chr = sub(".hdf", "", sub(".*_OKPR_", "", file))
+    ) |>
+    dplyr::mutate(
+      time_pos = strptime(time_chr, "%Y%m%d%H%M%S", tz = "UTC")
```
[the remark this answers is not in the document]
Comment (bart1) [reply R/get_pvol_cz.R:35]: I guess I'm so familiar with `strptime` I never really considered the readability, what is the lubridate alternative?

File: R/get_pvol.R [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    !inherits(time, "POSIXct") ||
+    anyDuplicated(time) ||
+    any((as.numeric(time) %% 300) != 0)) {
+    cli_abort("The argument {.arg time} to the {.fn get_pvol} function should be a POSIXct without duplications. All timestamps should be rounded to 5 minutes intervals.",
+      class = "getRad_error_time_not_correct"
+    )
+  }
+  if (length(time) != 1) {
+    l <- (purrr::map(time, get_pvol, radar = radar, ...))
+    if (length(radar) != 1) {
```
[the remark this answers is not in the document]
Comment (bart1) [reply R/get_pvol.R:42]: The alternative would be to first expand the combination of radars and time and then loop over that but that is not really easier it just makes the function slightly less recursive.

**Comment by bart1** [comment]
The Estonian test does not seem to fail much so maybe that was temporary, I have enabled them

File: R/get_pvol_cz.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (bart1) [reply R/get_pvol_cz.R:9]: Actually currently `bioRad` currently adds most reverse dependencies. If we can drop that in de future the package becomes much lighter. As currently pretty much all functions use it  (only not Germany). for now that would not be very useful, but good to keep in mind.

**Comment by PietrH** [comment]
> Can you list your design notes as issues? Easier to answer and track.
> 
> Longer term we’ll have to be careful not to introduce circular dependencies between this and bioRad

**Comment by PietrH** [comment]
> Maybe a few more notes on some design:
> 
>     * I try to consistently set the user agent so requests through the package can be tracked. Would it be useful to add the package version number or commit here? If so how to keep that up to date.
> 
>     * I now class each error as `"getRad_error_...."` we could also go for `c("getRad_error_....","getRad_error")` more like `httr2`
> 
>     * currently `time` should be a `POSIXct` rounded to 5 minutes we might consider also allowing for intervals and multiple timestamps. The individual functions now only download one file at a time and do not parallelize requests.

Are these all issues/implemented?

**Comment by PietrH** [comment]
> > Sure, but we still have to keep description up to date. To help with this we might want to setup an extra check in CI
> > […](#)
> > On Fri, 8 Nov 2024, 16:42 bart1, _**@**_._**> wrote: _**@**_.**_* commented on this pull request. ------------------------------ In R/aaa.R <[#1 (comment)](https://github.com/enram/getRad/pull/1#discussion_r1834604450)> : > @@ -0,0 +1,18 @@ +.onLoad <- function(libname, pkgname) { # nolint + + op <- options() + op.getRad <- list( + getRad.key_prefix = "getRad_", + getRad.user_agent = paste("R package getRad", getNamespaceVersion("getRad")) I think getNamespaceVersion does not get outdated right? It just grabs from the Description/installation — Reply to this email directly, view it on GitHub
> 
> I think it will keep roughly up to date, when I checked with peter he did not know a way to add the git hash which could be an improvement but otherwise I think it is good enough for the time being

If you were to add the git hash, wouldn't that be the hash of the previous commit relative to the version? That seems a bit confusing to me, is there any way to get the git hash without committing it? 

I like the idea in principle, but I also don't know how to implement it. Let's stick with more traditional version numbers for now.

**Comment by bart1** [comment]
> * I try to consistently set the user agent so requests through the package can be tracked. Would it be useful to add the package version number or commit here? If so how to keep that up to date.
Done through `getNamespaceVersion`
> * I now class each error as `"getRad_error_...."` we could also go for `c("getRad_error_....","getRad_error")` more like `httr2`
Decided to keep as is for the time being
> * currently `time` should be a `POSIXct` rounded to 5 minutes we might consider also allowing for intervals and multiple timestamps. The individual functions now only download one file at a time and do not parallelize requests.
For now it is good in the future we might generalize see #2

**Review by PietrH — APPROVED** [review]

**Pull request merged**